src/peergos/server/space/JdbcUsageStore.java

```diff
@@ -0,0 +1,370 @@
+                "WHERE writer_id = ?;")) {
+            insert.setLong(1, 0);
+            insert.setInt(2, writerId);
+            int count = insert.executeUpdate();
+            if (count != 1)
+                throw new IllegalStateException("Didn't update one record!");
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    private int getPending(int writerId) {
+        try (PreparedStatement insert = conn.prepareStatement("SELECT pending_bytes FROM pendingusage WHERE writer_id = ?;")) {
+            insert.setInt(1, writerId);
+            ResultSet resultSet = insert.executeQuery();
+            resultSet.next();
+            return resultSet.getInt(1);
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    @Override
+    public void addPendingUsage(String username, PublicKeyHash writer, int size) {
+        int userId = getUserId(username);
+        int writerId = getWriterId(writer);
+        try (PreparedStatement insert = conn.prepareStatement(commands.insertOrIgnoreCommand("INSERT ", "INTO pendingusage (user_id, writer_id, pending_bytes) VALUES(?, ?, ?)"))) {
+            insert.setInt(1, userId);
+            insert.setInt(2, writerId);
+            insert.setInt(3, 0);
+            insert.executeUpdate();
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+        try (PreparedStatement insert = conn.prepareStatement("UPDATE pendingusage SET pending_bytes = pending_bytes + ? " +
+                "WHERE writer_id = ?;")) {
+            insert.setLong(1, size);
+            insert.setInt(2, writerId);
+            int count = insert.executeUpdate();
+            if (count != 1)
+                throw new IllegalStateException("Didn't update one record!");
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    @Override
+    public void setErrored(boolean errored, String username, PublicKeyHash writer) {
+        int userId = getUserId(username);
+        try (PreparedStatement insert = conn.prepareStatement("UPDATE userusage SET errored = ? " +
+                "WHERE user_id = ?;")) {
+            insert.setBoolean(1, errored);
+            insert.setInt(3, userId);
+            boolean updated1Record = insert.execute();
+            if (! updated1Record)
+                throw new IllegalStateException("Didn't update one record!");
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    @Override
+    public UserUsage getUsage(String username) {
```

Comment:
Ths should be one sql statement where there is a join between userusage and pendingusage

Reply:
Sounds good. How's your sql? :-)

Reply:
like this: SELECT pu.writer_id, pu.pending_bytes, uu.total_bytes, uu.errored FROM userusage uu, pendingusage pu WHERE uu.user_id = pu.user_id
AND uu.user_id = 1

Reply:
done, though the logic around total bytes and errored is ugly